### ISSUE TITLE
UserWarning: None of the inputs have requires_grad=True. Gradients will be None   warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")

### ISSUE THREAD

**Issue by liumingzhu6060** [eric-mitchell/direct-preference-optimization#9]
gradient_accumulation_steps=2，activation_checkpointing=true，FSDP train，
Can the following warnings be ignored？
UserWarning: None of the inputs have requires_grad=True. Gradients will be None   warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")

**Comment by liumingzhu6060** [comment]
gradient_accumulation_steps=2 is also this problem！

**Comment by liumingzhu6060** [comment]
gradient_accumulation_steps=1 is also same problem

**Comment by eric-mitchell** [comment]
Can you share the full command you launched? And ideally the wandb output?

**Comment by liumingzhu6060** [comment]
the full command: python -u train.py model=blank_model model.name_or_path=/home/liumingzhu/v0.2.4/ model.block_name=LlamaDecoderLayer datasets=[hh] loss=dpo loss.beta=0.1 exp_name=dpo_PCI_V_0_2_4 gradient_accumulation_steps=1 batch_size=8 eval_batch_size=8 trainer=FSDPTrainer sample_during_eval=false model.fsdp_policy_mp=bfloat16 activation_checkpointing=true

**Comment by liumingzhu6060** [comment]
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/79501943/879bbf6a-b52e-4c04-aaa4-6ba39b21d7bd)
I have added this code, as if the problem has been resolved

**Comment by liumingzhu6060** [comment]
the wandb output is this:
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/79501943/16f4220f-7099-4d1f-9b72-7c5a01d2248d)
it seams not normal!

**Comment by liumingzhu6060** [comment]
I add this code, The problem seems to have been resolved，but I need to check the results to confirm. 
Do you think this method is feasible？
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/79501943/978e6693-d63a-4e66-b6c2-47ec2f377308)

**Comment by liumingzhu6060** [comment]
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/79501943/3cdfbbd4-8c9b-46cc-93d4-b6e0de8d6cab)
When saving policy.pt, this problem occurs, do you konw why

**Comment by eric-mitchell** [comment]
Glad you figured out the checkpointing issue- could you share exactly what you changed, in case we should change the repository.

Separately, for your crash, I'd guess you might be running out of CPU RAM, but not sure. Feel free to open a separate issue for that question if you're still having problems with it.

**Comment by liumingzhu6060** [comment]
I changed my codes in trainers.py:
if config.activation_checkpointing:
    rank0_print('Attempting to enable activation checkpointing...')
    try:
        # use activation checkpointing, according to:
        # https://pytorch.org/blog/scaling-multimodal-foundation-models-in-torchmultimodal-with-pytorch-distributed/
        #
        # first, verify we have FSDP activation support ready by importing:
        from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
            checkpoint_wrapper,
            CheckpointImpl,
            apply_activation_checkpointing,
        )
    except Exception as e:
        rank0_print('FSDP activation checkpointing not available:', e)
    else:
        #################### added  ####################
        non_reentrant_wrapper = functools.partial(
            checkpoint_wrapper,
            offload_to_cpu=False,  
            checkpoint_impl=CheckpointImpl.NO_REENTRANT,
        )
        #################### added  ####################
        
        check_fn = lambda submodule: isinstance(submodule, wrap_class) 
        rank0_print('Applying activation checkpointing wrapper to policy...')
        # apply_activation_checkpointing(self.policy, checkpoint_wrapper_fn=checkpoint_wrapper, check_fn=check_fn)  # privious
        apply_activation_checkpointing(self.policy, checkpoint_wrapper_fn=non_reentrant_wrapper, check_fn=check_fn)  # added
        rank0_print('FSDP activation checkpointing enabled!')

**Comment by eric-mitchell** [comment]
Fixed by https://github.com/eric-mitchell/direct-preference-optimization/pull/16.

@liumingzhu6060 thank you for sharing your fix! If you're still having issues saving checkpoints/OOM, feel free to open a separate issue.
